### PR TITLE
Support options(environment variable) to enable grpc reuse port.

### PR DESCRIPTION
**Pull request by liutongxuan**
ReusePort scenario: parent process occupies the port, then share
the port through service such as ZooKeeper, and then child
process (TensorFlow process) reuse the port.

### CODE REVIEW

**Review by mihaimaruseac — DISMISSED** [review]

**Comment by gbaned** [comment]
@liutongxuan Can you please fix build failures ? Thanks!

**Comment by liutongxuan** [comment]
> @liutongxuan Can you please fix build failures ? Thanks!

fixed, thx.

**Review by mihaimaruseac — APPROVED** [review]

**Review by mihaimaruseac — DISMISSED** [review]

**Review by haoyuz — CHANGES_REQUESTED** [review]

File: tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc [hunk 1]
```diff
@@ -220,8 +232,14 @@ Status GrpcServer::Init(const GrpcServerOptions& opts) {
                            GetServerCredentials(server_def_), &bound_port_);
   builder.SetMaxMessageSize(std::numeric_limits<int32>::max());
 
-  builder.SetOption(
-      std::unique_ptr<::grpc::ServerBuilderOption>(new NoReusePortOption));
+  bool reuse_port = false;
+  ReadBoolFromEnvVar("TF_GRPC_REUSE_PORT", false, &reuse_port)
+      .IgnoreError();
```
[the remark this answers is not in the document]
Comment (liutongxuan) [reply tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc:237]: thx, done

**Review by mihaimaruseac — DISMISSED** [review]

**Review by mihaimaruseac — APPROVED** [review]

**Comment by burgerkingeater** [comment]
@liutongxuan have you tried enabling this flag and run a job with multi worker mirrored strategy? Mine got suck in `tensorflow::GrpcRemoteMaster::RunStep`:

```
#0  0x00007f579fa6e483 in epoll_wait () from /lib64/libc.so.6
#1  0x00007f576b1eb0d0 in pollset_work ()
   from /grid/b/tmp/yarn/usercache/chren/appcache/application_1599079061958_19332/container_e54_1599079061958_19332_01_000002/tf-edge-azkaban_3cc86d1d69b5f2c657cdd2eeb09ef3c50c9f336187f5edbef17623efa53cb0ef/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#2  0x00007f576b20c825 in cq_pluck(grpc_completion_queue*, void*, gpr_timespec, void*) ()
   from /grid/b/tmp/yarn/usercache/chren/appcache/application_1599079061958_19332/container_e54_1599079061958_19332_01_000002/tf-edge-azkaban_3cc86d1d69b5f2c657cdd2eeb09ef3c50c9f336187f5edbef17623efa53cb0ef/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#3  0x00007f5767caf6e2 in grpc_impl::CompletionQueue::Pluck(grpc::internal::CompletionQueueTag*) ()
   from /grid/b/tmp/yarn/usercache/chren/appcache/application_1599079061958_19332/container_e54_1599079061958_19332_01_000002/tf-edge-azkaban_3cc86d1d69b5f2c657cdd2eeb09ef3c50c9f336187f5edbef17623efa53cb0ef/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#4  0x00007f5767cb5087 in grpc::internal::BlockingUnaryCallImpl<tensorflow::RunStepRequest, tensorflow::RunStepResponse>::BlockingUnaryCallImpl(grpc::ChannelInterface*, grpc::internal::RpcMethod const&, grpc_impl::ClientContext*, tensorflow::RunStepRequest const&, tensorflow::RunStepResponse*) ()
   from /grid/b/tmp/yarn/usercache/chren/appcache/application_1599079061958_19332/container_e54_1599079061958_19332_01_000002/tf-edge-azkaban_3cc86d1d69b5f2c657cdd2eeb09ef3c50c9f336187f5edbef17623efa53cb0ef/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#5  0x00007f5767cb52e0 in tensorflow::grpc::MasterService::Stub::RunStep(grpc_impl::ClientContext*, tensorflow::RunStepRequest const&, tensorflow::RunStepResponse*) ()
   from /grid/b/tmp/yarn/usercache/chren/appcache/application_1599079061958_19332/container_e54_1599079061958_19332_01_000002/tf-edge-azkaban_3cc86d1d69b5f2c657cdd2eeb09ef3c50c9f336187f5edbef17623efa53cb0ef/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#6  0x00007f57670e651b in tensorflow::Status tensorflow::GrpcRemoteMaster::CallWithRetry<tensorflow::RunStepRequest, tensorflow::RunStepResponse>(tensorflow::CallOptions*, tensorflow::RunStepRequest const*, tensorflow::RunStepResponse*, grpc::Status (tensorflow::grpc::MasterService::Stub::*)(grpc_impl::ClientContext*, tensorflow::RunStepRequest const&, tensorflow::RunStepResponse*), std::string) ()
   from /grid/b/tmp/yarn/usercache/chren/appcache/application_1599079061958_19332/container_e54_1599079061958_19332_01_000002/tf-edge-azkaban_3cc86d1d69b5f2c657cdd2eeb09ef3c50c9f336187f5edbef17623efa53cb0ef/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#7  0x00007f57670e744d in tensorflow::GrpcRemoteMaster::RunStep(tensorflow::CallOptions*, tensorflow::RunStepRequestWrapper*, tensorflow::MutableRunStepResponseWrapper*) ()
   from /grid/b/tmp/yarn/usercache/chren/appcache/application_1599079061958_19332/container_e54_1599079061958_19332_01_000002/tf-edge-azkaban_3cc86d1d69b5f2c657cdd2eeb09ef3c50c9f336187f5edbef17623efa53cb0ef/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#8  0x00007f57670db077 in tensorflow::GrpcSession::RunProto(tensorflow::CallOptions*, tensorflow::MutableRunStepRequestWrapper*, tensorflow::MutableRunStepResponseWrapper*) ()
   from /grid/b/tmp/yarn/usercache/chren/appcache/application_1599079061958_19332/container_e54_1599079061958_19332_01_000002/tf-edge-azkaban_3cc86d1d69b5f2c657cdd2eeb09ef3c50c9f336187f5edbef17623efa53cb0ef/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#9  0x00007f57670dbdae in tensorflow::GrpcSession::RunHelper(tensorflow::RunOptions const&, std::vector<std::pair<std::string, tensorflow::Tensor>, std::allocator<std::pair<std::string, tensorflow::Tensor> > > const&, std::vector<std::string, std::allocator<std::string> > const&, std::vector<std::string, std::allocator<std::string> > const&, std::vector<tensorflow::Tensor, std::allocator<tensorflow::Tensor> >*, tensorflow::RunMetadata*, std::string const&) ()
   from /grid/b/tmp/yarn/usercache/chren/appcache/application_1599079061958_19332/container_e54_1599079061958_19332_01_000002/tf-edge-azkaban_3cc86d1d69b5f2c657cdd2eeb09ef3c50c9f336187f5edbef17623efa53cb0ef/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#10 0x00007f57670dc3d1 in tensorflow::GrpcSession::Run(tensorflow::RunOptions const&, std::vector<std::pair<std::string, tensorflow::Tensor>, std::allocator<std::pair<std::string, tensorflow::Tensor> > > const&, std::vector<std::string, std::allocator<std::string> > const&, std::vector<std::string, std::allocator<std::string> > const&, std::vector<tensorflow::Tensor, std::allocator<tensorflow::Tensor> >*, tensorflow::RunMetadata*) ()
   from /grid/b/tmp/yarn/usercache/chren/appcache/application_1599079061958_19332/container_e54_1599079061958_19332_01_000002/tf-edge-azkaban_3cc86d1d69b5f2c657cdd2eeb09ef3c50c9f336187f5edbef17623efa53cb0ef/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#11 0x00007f57670b4692 in tensorflow::SessionRef::Run(tensorflow::RunOptions const&, std::vector<std::pair<std::string, tensorflow::Tensor>, std::allocator<std::pair<std::string, tensorflow::Tensor> > > const&, std::vector<std::string, std::allocator<std::string> > const&, std::vector<std::string, std::allocator<std::string> > const&, std::vector<tensorflow::Tensor, std::allocator<tensorflow::Tensor> >*, tensorflow::RunMetadata*) ()
   from /grid/b/tmp/yarn/usercache/chren/appcache/application_1599079061958_19332/container_e54_1599079061958_19332_01_000002/tf-edge-azkaban_3cc86d1d69b5f2c657cdd2eeb09ef3c50c9f336187f5edbef17623efa53cb0ef/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#12 0x00007f576793bcc2 in TF_Run_Helper(tensorflow::Session*, char const*, TF_Buffer const*, std::vector<std::pair<std::string, tensorflow::Tensor>, std::allocator<std::pair<std::string, tensorflow::Tensor> > > const&, std::vector<std::string, std::allocator<std::string> > const&, TF_Tensor**, std::vector<std::string, std::allocator<std::string> > const&, TF_Buffer*, TF_Status*) ()
   from /grid/b/tmp/yarn/usercache/chren/appcache/application_1599079061958_19332/container_e54_1599079061958_19332_01_000002/tf-edge-azkaban_3cc86d1d69b5f2c657cdd2eeb09ef3c50c9f336187f5edbef17623efa53cb0ef/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#13 0x00007f576793c92a in TF_SessionRun ()
   from /grid/b/tmp/yarn/usercache/chren/appcache/application_1599079061958_19332/container_e54_1599079061958_19332_01_000002/tf-edge-azkaban_3cc86d1d69b5f2c657cdd2eeb09ef3c50c9f336187f5edbef17623efa53cb0ef/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#14 0x00007f57670b1024 in tensorflow::TF_SessionRun_wrapper_helper(TF_Session*, char const*, TF_Buffer const*, std::vector<TF_Output, std::allocator<TF_Output> > const&, std::vector<_object*, std::allocator<_---Type <return> to continue, or q <return> to quit---
object*> > const&, std::vector<TF_Output, std::allocator<TF_Output> > const&, std::vector<TF_Operation*, std::allocator<TF_Operation*> > const&, TF_Buffer*, TF_Status*, std::vector<_object*, std::allocator<_object*> >*) ()
   from /grid/b/tmp/yarn/usercache/chren/appcache/application_1599079061958_19332/container_e54_1599079061958_19332_01_000002/tf-edge-azkaban_3cc86d1d69b5f2c657cdd2eeb09ef3c50c9f336187f5edbef17623efa53cb0ef/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#15 0x00007f57670b11b2 in tensorflow::TF_SessionRun_wrapper(TF_Session*, TF_Buffer const*, std::vector<TF_Output, std::allocator<TF_Output> > const&, std::vector<_object*, std::allocator<_object*> > const&, std::vector<TF_Output, std::allocator<TF_Output> > const&, std::vector<TF_Operation*, std::allocator<TF_Operation*> > const&, TF_Buffer*, TF_Status*, std::vector<_object*, std::allocator<_object*> >*) ()
   from /grid/b/tmp/yarn/usercache/chren/appcache/application_1599079061958_19332/container_e54_1599079061958_19332_01_000002/tf-edge-azkaban_3cc86d1d69b5f2c657cdd2eeb09ef3c50c9f336187f5edbef17623efa53cb0ef/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#16 0x00007f576100d9c3 in void pybind11::cpp_function::initialize<pybind11_init__pywrap_tf_session(pybind11::module&)::{lambda(TF_Session*, TF_Buffer*, pybind11::handle const&, std::vector<TF_Output, std::allocator<TF_Output> > const&, std::vector<TF_Operation*, std::allocator<TF_Operation*> > const&, TF_Buffer*)#17}, pybind11::object, TF_Session*, TF_Buffer*, pybind11::handle const&, std::vector<TF_Output, std::allocator<TF_Output> > const&, std::vector<TF_Operation*, std::allocator<TF_Operation*> > const&, TF_Buffer*, pybind11::name, pybind11::scope, pybind11::sibling>(pybind11_init__pywrap_tf_session(pybind11::module&)::{lambda(TF_Session*, TF_Buffer*, pybind11::handle const&, std::vector<TF_Output, std::allocator<TF_Output> > const&, std::vector<TF_Operation*, std::allocator<TF_Operation*> > const&, TF_Buffer*)#17}&&, pybind11::object (*)(TF_Session*, TF_Buffer*, pybind11::handle const&, std::vector<TF_Output, std::allocator<TF_Output> > const&, std::vector<TF_Operation*, std::allocator<TF_Operation*> > const&, TF_Buffer*), pybind11::name const&, pybind11::scope const&, pybind11::sibling const&)::{lambda(pybind11::detail::function_call&)#3}::_FUN(pybind11::detail::function_call) ()
   from /grid/b/tmp/yarn/usercache/chren/appcache/application_1599079061958_19332/container_e54_1599079061958_19332_01_000002/tf-edge-azkaban_3cc86d1d69b5f2c657cdd2eeb09ef3c50c9f336187f5edbef17623efa53cb0ef/site-packages/tensorflow/python/_pywrap_tf_session.so
#17 0x00007f5760ff9cc0 in pybind11::cpp_function::dispatcher(_object*, _object*, _object*) ()
   from /grid/b/tmp/yarn/usercache/chren/appcache/application_1599079061958_19332/container_e54_1599079061958_19332_01_000002/tf-edge-azkaban_3cc86d1d69b5f2c657cdd2eeb09ef3c50c9f336187f5edbef17623efa53cb0ef/site-packages/tensorflow/python/_pywrap_tf_session.so
#18 0x00007f57a093049b in _PyMethodDef_RawFastCallKeywords () from /export/apps/python/3.7.7/lib/libpython3.7m.so.1.0
```

**Pull request merged**